bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java

```diff
@@ -0,0 +1,298 @@
+		clientSocket.close();
+	}
+
+	/**
+	 * @{inheritDoc
+	 */
+	@Override
+	protected long getRefreshInterval() {
+		return refreshInterval;
+	}
+
+	/**
+	 * @{inheritDoc
+	 */
+	@Override
+	protected String getName() {
+		return "TACmi Refresh Service";
+	}
+
+	/**
+	 * @{inheritDoc
+	 */
+	@Override
+	protected void execute() {
+		logger.debug("execute() method is called!");
+		try {
+			clientSocket.setBroadcast(true);
+			clientSocket.setSoTimeout(10000);
+			byte[] receiveData = new byte[14];
+			DatagramPacket receivePacket = new DatagramPacket(receiveData,
+					receiveData.length);
+			clientSocket.receive(receivePacket);
+			byte[] data = receivePacket.getData();
+			if (data[1] > 0) {
+				logger.debug("Processing analog message");
+				AnalogMessage message = new AnalogMessage(data);
+				logger.debug(message.toString());
+				for (TACmiBindingProvider provider : providers) {
+					for (String itemName : provider.getItemNames()) {
+						logger.debug("Processing item: " + itemName);
+						if (provider.getCanNode(itemName) == message.canNode
+								&& provider.getPortType(itemName).equals("a")) {
+							if (message.hasPortnumber(provider
+									.getPortNumber(itemName))) {
+								AnalogValue value = message
+										.getAnalogValue(provider
+												.getPortNumber(itemName));
+								if (value.value != null) {
+									logger.debug("Updating item: " + itemName
+											+ " with value: " + value.value);
+									eventPublisher.postUpdate(itemName,
+											new DecimalType(value.value));
+								} else {
+									logger.debug(
+											"Portnumber {} not included in message",
+											provider.getPortNumber(itemName));
+								}
+							}
+						} else {
+							logger.debug("CAN Node does not match or item not analog");
+						}
+					}
+				}
+			} else if (data[1] == 0) {
+				logger.debug("Processing digital message");
+				DigitalMessage message = new DigitalMessage(data);
+				logger.debug(message.toString());
+				for (TACmiBindingProvider provider : providers) {
+					for (String itemName : provider.getItemNames()) {
+						logger.debug("Processing item: " + itemName);
+						if (provider.getCanNode(itemName) == message.canNode
+								&& provider.getPortType(itemName).equals("d")) {
+							if (message.getPortState(provider
+									.getPortNumber(itemName))) {
+								logger.debug("Updating item {} with state {}",
+										itemName, message.getPortState(provider
+												.getPortNumber(itemName)));
+								eventPublisher.postUpdate(itemName,
+										OnOffType.ON);
+							} else {
+								logger.debug("Updating item {} with state {}",
+										itemName, message.getPortState(provider
+												.getPortNumber(itemName)));
+								eventPublisher.postUpdate(itemName,
+										OnOffType.OFF);
+							}
+						} else {
+							logger.debug(
+									"CAN Node {} does not match or item not digital",
+									message.canNode);
+						}
+					}
+				}
+			}
```

Comment:
there seem to be quite a lot code duplication in ```if``` and ```else if``` when moving the loops outside the if-else the duplication could be reduced (if not avoided)